### PR TITLE
Simplify JVP of `lax.div` by reusing already computed values

### PR DESCRIPTION
**Pull request by lgeiger**
This simplifies the JVP of `lax.div` by reusing its result $r = \frac{x}{y}$, since $\frac{\partial r}{\partial y} = - \frac{x}{y^2} = - \frac{r}{y}$.

The change improves the generated HLO which might help with downstream compiler passes. Click expand to see the before and after of the HLO generated by
```python
jax.value_and_grad(lax.div, argnums=1)(jnp.ones(()), jnp.ones(()))
```

<details><summary>HLO before (click to expand)</summary>

```mlir
module @jit_div attributes {mhlo.num_partitions = 1 : i32, mhlo.num_replicas = 1 : i32} {
  func.func public @main(%arg0: tensor<f32> {jax.arg_info = "x", mhlo.sharding = "{replicated}"}, %arg1: tensor<f32> {jax.arg_info = "y", mhlo.sharding = "{replicated}"}) -> (tensor<f32> {jax.result_info = "[0]"}, tensor<f32> {jax.result_info = "[1]"}) {
    %0 = stablehlo.divide %arg0, %arg1 : tensor<f32>
    %1 = stablehlo.multiply %arg1, %arg1 : tensor<f32>
    %2 = stablehlo.constant dense<1.000000e+00> : tensor<f32>
    %3 = stablehlo.divide %2, %1 : tensor<f32>
    %4 = stablehlo.constant dense<1.000000e+00> : tensor<f32>
    %5 = stablehlo.multiply %4, %3 : tensor<f32>
    %6 = stablehlo.multiply %5, %arg0 : tensor<f32>
    %7 = stablehlo.negate %6 : tensor<f32>
    return %0, %7 : tensor<f32>, tensor<f32>
  }
}
```

</details> 

<details><summary>HLO after (click to expand)</summary>

```mlir
module @jit_div attributes {mhlo.num_partitions = 1 : i32, mhlo.num_replicas = 1 : i32} {
  func.func public @main(%arg0: tensor<f32> {jax.arg_info = "x", mhlo.sharding = "{replicated}"}, %arg1: tensor<f32> {jax.arg_info = "y", mhlo.sharding = "{replicated}"}) -> (tensor<f32> {jax.result_info = "[0]"}, tensor<f32> {jax.result_info = "[1]"}) {
    %0 = stablehlo.divide %arg0, %arg1 : tensor<f32>
    %1 = stablehlo.divide %0, %arg1 : tensor<f32>
    %2 = stablehlo.constant dense<1.000000e+00> : tensor<f32>
    %3 = stablehlo.multiply %2, %1 : tensor<f32>
    %4 = stablehlo.negate %3 : tensor<f32>
    return %0, %4 : tensor<f32>, tensor<f32>
  }
}
```

</details>

### CODE REVIEW

**Comment by jakevdp** [comment]
Looks good in principle, but this kind of change can have subtle downstream effects. Looping in @mattjj who has good intuition for this kind of stuff.

**Comment by lgeiger** [comment]
> This looks good in principle, but this kind of change can have subtle downstream effects.

Indeed, I just noticed that `testSoftmaxGradFlag` is failing since the size of `ad_checkpoint.saved_residuals(nn.softmax, x)`  inside `jax.softmax_custom_jvp(False)` changed. I'm happy to close this PR if this change is risky as the potential benefits are not really that large.

**Comment by mattjj** [comment]
That particular test is brand new so don't worry about it yet. The main intent there was to check something else (though we should still understand what's changing here in terms of residuals saved).

Actually the main thing I want to do is check the history of this rule to see if there was any reason to write it a particular way.

**Comment by lgeiger** [comment]
> Actually the main thing I want to do is check the history of this rule to see if there was any reason to write it a particular way.

I checked git blame and it seems that jvp started out as https://github.com/google/jax/blob/17e5783e5f0fb58c12194cf0e556abb7f1ac438d/jax/lax/lax.py#L1721-L1723 and got changed in #1388 (`pow` -> `square`) and #3140 (`square` to `integer_pow`).
Looking at the last PR in more detail it seems like my change could indeed have downsides for higher order gradients as
```python
jax.jit(grad(grad(grad(lambda x: 1/x)))).lower(3.)
```
seems to generate a much larger HLO module compared to before which could become a problem in practice.
